arbos/l1pricing/l1pricing.go

```diff
@@ -5,242 +5,358 @@ package l1pricing
+func (ps *L1PricingState) FundsDueToSequencer() (*big.Int, error) {
+	return ps.fundsDueToSequencer.Get()
+}
+
+func (ps *L1PricingState) SetFundsDueToSequencer(amt *big.Int) error {
+	return ps.fundsDueToSequencer.Set(amt)
+}
+
+func (ps *L1PricingState) FundsDueForRewards() (*big.Int, error) {
+	return ps.fundsDueForRewards.Get()
+}
+
+func (ps *L1PricingState) SetFundsDueForRewards(amt *big.Int) error {
+	return ps.fundsDueForRewards.Set(amt)
+}
+
+func (ps *L1PricingState) UnitsSinceUpdate() (uint64, error) {
+	return ps.unitsSinceUpdate.Get()
 }
 
-// Get the aggregator who is eligible to be reimbursed for L1 costs of txs from sender, or nil if there is none.
-func (ps *L1PricingState) ReimbursableAggregatorForSender(sender common.Address) (*common.Address, error) {
-	fromTable, err := ps.UserSpecifiedAggregator(sender)
+func (ps *L1PricingState) SetUnitsSinceUpdate(units uint64) error {
+	return ps.unitsSinceUpdate.Set(units)
+}
+
+func (ps *L1PricingState) PricePerUnit() (*big.Int, error) {
+	return ps.pricePerUnit.Get()
+}
+
+func (ps *L1PricingState) SetPricePerUnit(price *big.Int) error {
+	return ps.pricePerUnit.Set(price)
+}
+
+// Update the pricing model with info from the start of a block
+func (ps *L1PricingState) UpdateTime(currentTime uint64) {
+	_ = ps.SetCurrentTime(currentTime)
+}
+
+// Update the pricing model based on a payment by the sequencer
+func (ps *L1PricingState) UpdateForSequencerSpending(stateDb vm.StateDB, updateTime uint64, currentTime uint64, weiSpent *big.Int) error {
+	// compute previous shortfall
+	fundsDueToSequencer, err := ps.FundsDueToSequencer()
 	if err != nil {
-		return nil, err
+		return err
 	}
-	if fromTable != nil {
-		return fromTable, nil
+	fundsDueForRewards, err := ps.FundsDueForRewards()
+	if err != nil {
+		return err
+	}
+	availableFunds, err := ps.AvailableFunds()
+	if err != nil {
+		return err
 	}
+	oldShortfall := new(big.Int).Sub(new(big.Int).Add(fundsDueToSequencer, fundsDueForRewards), availableFunds)
 
-	refuses, err := ps.RefusesDefaultAggregator(sender)
-	if err != nil || refuses {
-		return nil, err
+	// compute allocation fraction
+	lastUpdateTime, err := ps.LastUpdateTime()
+	if err != nil {
+		return err
 	}
-	aggregator, err := ps.DefaultAggregator()
+	if updateTime > currentTime || updateTime < lastUpdateTime || currentTime == lastUpdateTime {
+		return ErrInvalidTime
+	}
+	allocFractionNum := big.NewInt(int64(updateTime - lastUpdateTime))
+	allocFractionDenom := big.NewInt(int64(currentTime - lastUpdateTime))
+
+	// allocate units to this update
+	unitsSinceUpdate, err := ps.UnitsSinceUpdate()
 	if err != nil {
-		return nil, err
+		return err
 	}
-	if aggregator == (common.Address{}) {
-		return nil, nil
+	unitsAllocated := unitsSinceUpdate * (updateTime - lastUpdateTime) / (currentTime - lastUpdateTime)
+	unitsSinceUpdate -= unitsAllocated
+	if err := ps.SetUnitsSinceUpdate(unitsSinceUpdate); err != nil {
```

Comment:
This seems to only ever go down and there's no other callsites to `SetUnitsSinceUpdate`. We need to add units back in somewhere

Reply:
I'm guessing the idea was to only keep funds ephemerally in `L1PricerFundsPoolAddress`, and instead rely on `AvailableFunds` to determine how much we can mint. I'd propose instead keeping funds permanently in `L1PricerFundsPoolAddress` and exclusively transferring. Instead of `AvailableFunds`, we would track the balance of the `L1PricerFundsPoolAddress` at last update, and then use that to determine how much was collected this update.